api/odl.py

```diff
@@ -890,3 +956,509 @@ def _url_for(self, *args, **kwargs):
+            raise CannotHold()
+        entry = entries[0]
+        availability = entry.get("opds_availability", {})
+        if availability.get("status") not in ["reserved", "ready"]:
+            raise CannotHold()
+        start = datetime.datetime.strptime(availability.get("since"), self.TIME_FORMAT)
+        end = datetime.datetime.strptime(availability.get("until"), self.TIME_FORMAT)
+        position = entry.get("opds_holds", {}).get("position")
+        if position:
+            position = int(position)
+        # Get the loan base url from a link.
+        info_links = [link for link in entry.get("links") if link.get("rel") == "self"]
+        external_identifier = info_links[0].get("href")
+
+        return HoldInfo(
+            licensepool.collection,
+            licensepool.data_source.name,
+            licensepool.identifier.type,
+            licensepool.identifier.identifier,
+            start,
+            end,
+            hold_position=position,
+            external_identifier=external_identifier,
+        )
+
+    def release_hold(self, patron, pin, licensepool):
+        _db = Session.object_session(patron)
+
+        hold = get_one(
+            _db, Hold,
+            license_pool_id=licensepool.id,
+            patron=patron,
+        )
+        if not hold:
+            raise NotOnHold()
+
+        info_url = hold.external_identifier
+        try:
+            response = self._get(info_url)
+        except RemoteIntegrationException, e:
+            raise CannotReleaseHold()
+        feed = feedparser.parse(unicode(response.content))
+        entries = feed.get("entries")
+        if len(entries) < 1:
+            raise CannotReleaseHold()
+        entry = entries[0]
+        availability = entry.get("opds_availability", {})
+        if availability.get("status") not in ["reserved", "ready"]:
+            raise CannotReleaseHold()
+        revoke_links = [link for link in entry.get("links") if link.get("rel") == "http://librarysimplified.org/terms/rel/revoke"]
+        if len(revoke_links) < 1:
+            raise CannotReleaseHold()
+        revoke_url = revoke_links[0].get("href")
+        try:
+            self._get(revoke_url)
+        except RemoteIntegrationException, e:
+            raise CannotReleaseHold()
+        return True
+
+    def patron_activity(self, patron, pin):
+        _db = Session.object_session(patron)
+        loans = _db.query(Loan).join(Loan.license_pool).filter(
+            LicensePool.collection_id==self.collection_id
+        ).filter(
+            Loan.patron==patron
+        )
+
+        holds = _db.query(Hold).join(Hold.license_pool).filter(
+            LicensePool.collection_id==self.collection_id
+        ).filter(
+            Hold.patron==patron
+        )
+        
+        activity = []
+        for loan in loans:
+            info_url = loan.external_identifier
+            response = self._get(info_url, allowed_response_codes=["2xx", "3xx", "400"])
```

Comment:
Not much we can do at this point but this is a downside to 'every loan has a different patron identifier' that I hadn't previously noticed.

Reply:
For holds, it is likely that the hold position will change due to loans expiring or being returned on other circ managers. For loans, though, this would only be necessary if the loan was revoked through the DRM system without using the circ manager's revoke link, or was revoked by an admin on the ODL server.